### PR TITLE
LVNODCDXI-66

### PR DESCRIPTION
**Pull request by mr-redFox**
created Redis module for easy using
refactored server for more readability (split logic separated files auth controller and effects)
added a few new errors with status 4xx & 5xx for cleaner code

### CODE REVIEW

**Review by OstapDutkevych — DISMISSED** [review]

**Review by Kindest13 — APPROVED** [review]

File: src/server/controllers/authentication.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    const savedUser = await user.save();
+    if (savedUser) {
+      const tokens = await generateTokens({ userId: user._id });
+      return res.status(201).json({ token: tokens });
+    }
+  } catch (error) {
+    return response(res, error.message, 404);
+  }
+
+  return null;
```
Comment (tonypntsrn) [inline src/server/controllers/authentication.js]: I don’t return null in loginUser function, do you actually need it here?

File: src/server/controllers/authentication.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+};
+
+export const loginUser = async (req, res) => {
+  const { email, password } = req.body;
+
+  try {
+    const user = await User.findOne({ email });
+
+    if (!user) throw new ClientError('User not exist');
```
Comment (tonypntsrn) [inline src/server/controllers/authentication.js]: Please add curly bracers to this if statement here

File: src/server/controllers/authentication.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import { ClientError, TeapotError } from '../helpers/error';
+import { putInRedis, getFromRedis, deleteFromRedis } from '../helpers/redis';
+
+// internal helpers
+
+const generateTokens = async (user) => {
+  const tokens = generateTokensPair(user);
+  const result = await putInRedis(tokens.refresh, user.userId);
+
+  if (!result) throw new TeapotError();
```
Comment (tonypntsrn) [inline src/server/controllers/authentication.js]: Please, add curly bracers here

File: src/server/helpers/redis.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+redisClient.on('connect', () => process.env.DEV_MODE && console.log('redis connected '));
+
+/**
+ * return value (user id in this case) from redis storage finded by token (key)
+ * @param {string} key
+ * @return {Promise} with data (userId)
+ * @throws {ClientError}
+ */
+export function getFromRedis(key) {
+  // if ((typeof key) !== 'string') {
```
Comment (tonypntsrn) [inline src/server/helpers/redis.js]: Do you need this comments here?

File: src/server/helpers/redis.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+redisClient.on('error', err => console.log(`Redis error -> ${err}`));
+redisClient.on('connect', () => process.env.DEV_MODE && console.log('redis connected '));
+
+/**
+ * return value (user id in this case) from redis storage finded by token (key)
+ * @param {string} key
+ * @return {Promise} with data (userId)
+ * @throws {ClientError}
+ */
+export function getFromRedis(key) {
```
Comment (tonypntsrn) [inline src/server/helpers/redis.js]: Can it be rewrited using async/await?

File: src/server/helpers/redis.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * put token in redis storage key = refresh-token, value = userId
+ * this methodology can work with multiple sessions (multi device)
+ * @param {string} key
+ * @param {string} value
+ * @return {Promise} (boolean) true/false
+ * @throws {ClientError}
+ */
+export function putInRedis(key, value) {
```
Comment (tonypntsrn) [inline src/server/helpers/redis.js]: You are using almost same functions for put get, etc, could it be rewritten to just call one function and pass there methods as a parameters?

File: src/server/helpers/token.js [hunk 7]
```diff
@@ -6,7 +6,7 @@ import response from './errorHandler';
 dotenv.config();
 
 const { SECRET } = process.env;
-const tokenLife = '15m';
+const tokenLife = '5h';
```
Comment (tonypntsrn) [inline src/server/helpers/token.js]: Are you sure this is the secure life time for token?

**Review by Kindest13 — DISMISSED** [review]

File: src/server/controllers/authentication.js (hunk 1 above)
Comment (mr-redFox) [reply src/server/controllers/authentication.js]: fixed

File: src/server/controllers/authentication.js (hunk 2 above)
Comment (mr-redFox) [reply src/server/controllers/authentication.js]: added, done

File: src/server/controllers/authentication.js (hunk 3 above)
Comment (mr-redFox) [reply src/server/controllers/authentication.js]: added, done

File: src/server/helpers/redis.js (hunk 4 above)
Comment (mr-redFox) [reply src/server/helpers/redis.js]: fixed

File: src/server/helpers/redis.js (hunk 5 above)
Comment (mr-redFox) [reply src/server/helpers/redis.js]: done

File: src/server/helpers/redis.js (hunk 6 above)
Comment (mr-redFox) [reply src/server/helpers/redis.js]: rewrited used decorator for this logic

File: src/server/helpers/token.js (hunk 7 above)
Comment (mr-redFox) [reply src/server/helpers/token.js]: changed to 30m

File: src/server/controllers/effects.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+import response from '../helpers/errorHandler';
+import { NotFoundError, ClientError } from '../helpers/error';
+
+export const getEffect = async (req, res) => {
+  const { title } = req.body;
+  const { userId } = req;
+
+  try {
+    const effects = await User.findOne({ _id: userId }, 'effects');
+    const finded = effects.find(effect => effect.title === title);
```
Comment (tonypntsrn) [inline src/server/controllers/effects.js]: Funded what?
Names should be more explicit

File: src/server/controllers/authentication.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  const { email, password } = req.body;
+
+  try {
+    const user = await User.findOne({ email });
+
+    if (!user) {
+      throw new ClientError('User not exist');
+    }
+
+    const verified = await user.verifyPassword(password);
```
Comment (tonypntsrn) [inline src/server/controllers/authentication.js]: All names should be more explicit for all who reads code, for eg. I don’t understand what verified means

File: src/server/helpers/redis.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+const redisClient = redis.createClient({
+  url: process.env.REDIS_URL,
+  expire: DAY,
+});
+
+redisClient.on('error', err => console.log(`Redis error -> ${err}`));
+redisClient.on('connect', () => process.env.DEV_MODE && console.log('redis connected '));
+
+function injectMethod(method, methodArgs) {
```
Comment (tonypntsrn) [inline src/server/helpers/redis.js:23]: 👍

File: src/server/controllers/effects.js (hunk 8 above)
Comment (mr-redFox) [reply src/server/controllers/effects.js]: Done

File: src/server/controllers/authentication.js (hunk 9 above)
Comment (mr-redFox) [reply src/server/controllers/authentication.js]: Done

**Review by tonypntsrn — APPROVED** [review]

**Review by Kindest13 — APPROVED** [review]

**Pull request merged**